core/src/main/scala/caliban/execution/Executor.scala

```diff
@@ -101,46 +67,66 @@ object Executor {
             case _ => s
           }
-        case FunctionStep(step) => reduceStep(step(arguments), currentField, Map())
-        case ListStep(steps)    => reduceList(steps.map(reduceStep(_, currentField, arguments)))
+        case FunctionStep(step) => reduceStep(step(arguments), currentField, Map(), path)
+        case ListStep(steps) =>
+          reduceList(steps.zipWithIndex.map {
+            case (step, i) => reduceStep(step, currentField, arguments, Right(i) :: path)
+          })
         case ObjectStep(objectName, fields) =>
           val mergedFields = mergeFields(currentField, objectName)
           val items = mergedFields.map {
-            case Field(name @ "__typename", _, _, alias, _, _, _) =>
-              alias.getOrElse(name) -> PureStep(StringValue(objectName))
+            case f @ Field(name @ "__typename", _, _, alias, _, _, _) =>
+              (alias.getOrElse(name), PureStep(StringValue(objectName)), fieldInfo(f, path))
             case f @ Field(name, _, _, alias, _, _, args) =>
               val arguments = resolveVariables(args, variableDefinitions, variableValues)
-              alias.getOrElse(name) ->
+              (
+                alias.getOrElse(name),
                 fields
                   .get(name)
-                  .fold(NullStep: ReducedStep[R])(reduceStep(_, f, arguments))
+                  .fold(NullStep: ReducedStep[R])(reduceStep(_, f, arguments, Left(alias.getOrElse(name)) :: path)),
+                fieldInfo(f, path)
+              )
           }
-          reduceObject(items)
+          reduceObject(items, fieldWrappers)
         case QueryStep(inner) =>
           ReducedStep.QueryStep(
-            inner.bimap(
-              GenericSchema.effectfulExecutionError(currentField.name, _),
-              reduceStep(_, currentField, arguments)
-            )
+            inner.bimap(GenericSchema.effectfulExecutionError(path, _), reduceStep(_, currentField, arguments, path))
           )
         case StreamStep(stream) =>
           ReducedStep.StreamStep(
-            stream.bimap(
-              GenericSchema.effectfulExecutionError(currentField.name, _),
-              reduceStep(_, currentField, arguments)
-            )
+            stream.bimap(GenericSchema.effectfulExecutionError(path, _), reduceStep(_, currentField, arguments, path))
           )
       }
 
     def makeQuery(step: ReducedStep[R], errors: Ref[List[CalibanError]]): ZQuery[R, Nothing, ResponseValue] = {
+
+      def wrap(query: ZQuery[R, Nothing, ResponseValue])(
+        wrappers: List[FieldWrapper[R]],
+        fieldInfo: FieldInfo
+      ): ZQuery[R, Nothing, ResponseValue] =
+        wrappers match {
+          case Nil => query
+          case wrapper :: tail =>
+            ZQuery
+              .environment[R]
+              .flatMap(
+                env =>
+                  wrap(
+                    wrapper
+                      .f(query.provide(Described(env, "Wrapper")), fieldInfo)
```

Comment:
Maybe provide an optional name for the `Wrapper` when creating it? Otherwise won't all wrappers end up being equal to each other?

Reply:
Maybe use the same label syntax:

trait Wrapper[-R] {
  def ??(label: String): Wrapper[-R]
}

Reply:
Not really, the name required by `ZQuery` has to do with caching. `ZQuery` generates a hash to ensure the same request is queried only once and otherwise cached. When we modify the environment with `provide`, the request is the same but we need to change the hash to avoid reading from the same cache. It doesn't need to be different for each wrapper.

@adamgfraser can you confirm?

Reply:
Yes.

To be specific, internally we need to maintain a mapping from datasources to requests that are associated with those data sources. So if we transform the data source somehow, like by providing it with its environment, we need to modify the identifier for the data source so hash codes are consistent with equality.

We always modify the original identifier with the new descriptions. So if we had `DataSourceA` and `DataSourceB` and wrapped them each with three wrapped fields, we would get `DataSourceA.provide(Wrapper).provide(Wrapper).provide(Wrapper)` and `DataSourceB.provide(Wrapper).provide(Wrapper).provide(Wrapper)`, which would still be different.

The only situation where I could theoretically see it creating an issue is if you had the same data source and provided it with two different sets of the same number of wrappers that somehow operate on the data source, because then you could have two different data sources with the same identifier. But I don't think that is possible here because the `FieldWrapper` takes a `ZQuery[Any, _, _]` as an input, so I don't think it could even do anything to modify a data source.

Reply:
Could this be solved by changing the signature of `f` in `FieldWrapper` from:

```scala
(ZQuery[Any, Nothing, ResponseValue], FieldInfo) => ZQuery[R, CalibanError, ResponseValue]
```

to:

```scala
(ZQuery[R, Nothing, ResponseValue], FieldInfo) => ZQuery[R, CalibanError, ResponseValue]
```

That is, does the query being wrapper really have to have no environmental requirements at all? Or does it just not have to have any more requirements than the wrapper?

Reply:
I couldn’t get the variance to work if I kept R there, it was becoming covariant. I tried the technique with `R1 <: R` but still couldn’t make it compile.

Reply:
Could you make `FieldWrapper` invariant and change line 123 of `Wrapper.scala` to `wrapper: FieldWrapper[R]`?

Reply:
Oh right, I didn't realize I could make it invariant there while keeping it contravariant in `Wrapper`. I'll check if I can simplify the rest as well. Variance is tricky...

Reply:
@adamgfraser [really nice simplification](https://github.com/ghostdogpr/caliban/pull/165/commits/b60213667f7aa7891b5222c3240658f9b900711e), thank you! I no longer need `ZQuery.environment` but since it's here I just left it.